include/drivers/dai.h

```diff
@@ -0,0 +1,629 @@
+	 * memory block. This trigger can be used in RUNNING state only and at
+	 * first changes the interface state to STOPPING. When the current TX /
+	 * RX block is transmitted / received the state is changed to READY.
+	 * Subsequent START trigger will resume transmission / reception where
+	 * it stopped.
+	 */
+	DAI_TRIGGER_STOP,
+	/** @brief Optional - Post Stop the transmission / reception of data.
+	 *
+	 * Allows the DAI and downstream codecs to shutdown cleanly after audio
+	 * Tx/Rx by stopping any required clocks for downstream audio completion.
+	 */
+	DAI_TRIGGER_POST_STOP,
+	/** @brief Empty the transmit queue.
+	 *
+	 * Send all data in the transmit queue and stop the transmission.
+	 * If the trigger is applied to the RX queue it has the same effect as
+	 * DAI_TRIGGER_STOP. This trigger can be used in RUNNING state only and
+	 * at first changes the interface state to STOPPING. When all TX blocks
+	 * are transmitted the state is changed to READY.
+	 */
+	DAI_TRIGGER_DRAIN,
+	/** @brief Discard the transmit / receive queue.
+	 *
+	 * Stop the transmission / reception immediately and discard the
+	 * contents of the respective queue. This trigger can be used in any
+	 * state other than NOT_READY and changes the interface state to READY.
+	 */
+	DAI_TRIGGER_DROP,
+	/** @brief Prepare the queues after underrun/overrun error has occurred.
+	 *
+	 * This trigger can be used in ERROR state only and changes the
+	 * interface state to READY.
+	 */
+	DAI_TRIGGER_PREPARE,
+	/** @brief Reset
+	 *
+	 * This trigger frees resources and moves the driver back to initial
+	 * state as in driver init.
+	 */
+	DAI_TRIGGER_RESET,
+};
+
+/** \brief Properties of DAI
+ *
+ * This struct is used with APIs get_properties function to query DAI
+ * properties like fifo address and dma handshake. These are needed to
+ * for example to setup dma outside driver code.
+ */
+struct dai_properties {
+	uint32_t fifo_address; /* fifo address */
+	uint32_t dma_hs_id; /* dma handshake id */
+	uint32_t reg_init_delay; /* delay for register init */
+};
+
+/** Main dai config struct */
+struct dai_config {
+	enum dai_type type;
+	uint32_t dai_index;
+
+	/* generic dai config members */
+	uint8_t channels;
+	dai_fmt_t format;
+	dai_opt_t options;
+	uint32_t frame_clk_freq;
+
+	/* optional dai specific params */
+	const void *params;
+};
+
+/* TODO 1) array of clocks 2) wallclk naming to be clear */
+struct dai_ts_cfg {
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+	int type; /* SSP, DMIC, HDA, etc. */
+	int direction; /* Playback, capture */
+	int index; /* For SSPx to select correct timestamp register */
+	int dma_id; /* DMA instance id */
+	int dma_chan_index; /* Used DMA channel */
+	int dma_chan_count; /* Channels in single DMA */
+};
+
+struct dai_ts_data {
+	uint64_t walclk; /* Wall clock */
+	uint64_t sample; /* Sample count */
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
```

Comment:
It is rather _freq_  used in the SOF code, is _rate_ more common in Zephyr?

Reply:
sorry didn't get this.... these are actually direct copies from sof code. But can very well change to freq if better.

Reply:
I think we can close this in review in Zephyr upstream PR review. We might get guidance on timing data types that align better with Zephyr. To me, "walclk" seems a bit silly acronym.